ql/src/java/org/apache/hadoop/hive/ql/QueryCompilationBreakdownSummary.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hadoop.hive.ql.exec.tez.monitoring.PrintSummary;
+import org.apache.hadoop.hive.ql.log.PerfLogger;
+import org.apache.hadoop.hive.ql.session.SessionState;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.hadoop.hive.ql.exec.tez.monitoring.Constants.SEPARATOR;
+
+public class QueryCompilationBreakdownSummary implements PrintSummary {
```

Comment:
I just noticed that `PrintSummary` is a Tez interface so it doesn't seem appropriate for the use-case. It might have been part of my previous suggestion but I failed to notice then that it was tailored around Tez.

Thinking a bit more on the topic it seems that the `QueryCompilationBreakdownSummary` could be implemented as a `Hook`. In particular, the `QueryLifeTimeHook` interface seems to be fit well for this use case.

Two advantages of implementing this as a `QueryLifeTimeHook`:
1. we wouldn't need a new configuration property
2. we could easily test the summary feature via qtests

Reply:
I liked the idea of handling this as a Hook in general. Not sure if I completely understand the qtest part, can we automatically verify as part of the qtest out files or we need to look into the logs of the qfile test(which is what I did now) ?

Reply:
I think we can make the summary appear in the .q.out file. I added a few pointers in the review.